### PR TITLE
Use dynamically chosen ports to avoid address binding collisions

### PR DESCRIPTION
**Pull request by JunTaoLuo**
https://github.com/aspnet/Home/issues/3439 We should not be trying to bind to a static port for functional tests. Instead we should follow the pattern of binding to a dynamically determined port that's used everywhere else in our code base. I'll make similar changes to release/2.2 and master once this is approved.

### CODE REVIEW

**Comment by JunTaoLuo** [comment]
cc @muratg, this is a test code only to reduce test flakiness so there is no impact on product code.

**Comment by muratg** [comment]
Please wait for the branch to be open before checking in.

**Comment by JunTaoLuo** [comment]
FYI, I have made the fixes in release/2.2 and master with https://github.com/aspnet/WebSockets/commit/752f934a236e2e05b8ad985adfa540527643ae1a and https://github.com/aspnet/WebSockets/commit/899a8d733baeac11cb3996555f43bf910cb9d971. Since those branches are open and this PR is approved, I pushed directly.

**Comment by natemcmaster** [comment]
I didn't see this when I did my first pass of approved 2.1.5 issues. This repo needs a branding update, and Universe needs updating before this can merge. Please hold off for a bit. I'll merge this when I get those pieces in place.

**Comment by natemcmaster** [comment]
The PR checks for travis aren't passing for Linux or macOS. Can we fix this?

**Comment by anurse** [comment]
Looks like they're failing to install autobahn **again**. I'll take a look

**Comment by anurse** [comment]
Looks like the `-L` is missing from the `curl` command in `setup-wstest.sh` in this branch for some reason. I added a commit to fix it and introduced a hash check as well just to make sure it's working.

**Comment by JunTaoLuo** [comment]
I'm not as concerned about the travis tests but it seems like the osx one is still failing: `./build/setup-wstest.sh: line 24: sha256sum: command not found`

**Comment by anurse** [comment]
Yep, fixed. Guess macOS just isn't as cool and secure as Linux ;)

**Comment by JunTaoLuo** [comment]
`./build/setup-wstest.sh: line 24: sha1sum: command not found` I think travis might be running on an older osx version. I'm also okay with removing the sha check if necessary.

**Comment by JunTaoLuo** [comment]
Hmm the websocket tests are still failing on OSX...

**Comment by anurse** [comment]
```
Case Kestrel|SSL|ManagedSockets:1.1.7. Expected 'NON-STRICT' or 'OK', but got 'FAILED'
```

One of the autobahn tests failed. I don't see how your change could cause it but the macOS build on Travis is passing in the main branch.

**Comment by JunTaoLuo** [comment]
I don't see any passing builds on Travis for the 2.1 branch. Note that this is for the 2.1 branch, not master or 2.2. I think these failures may have always been there, but were masked by the setup-wstest.sh errors.

File: test/Microsoft.AspNetCore.WebSockets.Test/KestrelWebSocketHelpers.cs [hunk 1]
```diff
@@ -14,7 +14,7 @@ namespace Microsoft.AspNetCore.WebSockets.Test
 {
     public class KestrelWebSocketHelpers
     {
-        public static IDisposable CreateServer(ILoggerFactory loggerFactory, Func<HttpContext, Task> app)
+        public static IDisposable CreateServer(ILoggerFactory loggerFactory, Func<HttpContext, Task> app, int clientPort)
```
Comment (Tratcher) [inline test/Microsoft.AspNetCore.WebSockets.Test/KestrelWebSocketHelpers.cs]: clientPort -> port

File: test/Microsoft.AspNetCore.WebSockets.Test/WebSocketMiddlewareTests.cs [hunk 2]
```diff
@@ -23,10 +24,13 @@ namespace Microsoft.AspNetCore.WebSockets.Test
 #endif
     public class WebSocketMiddlewareTests : LoggedTest
     {
-        private static string ClientAddress = "ws://localhost:54321/";
+        private readonly int ClientPort;
```
Comment (Tratcher) [inline test/Microsoft.AspNetCore.WebSockets.Test/WebSocketMiddlewareTests.cs]: ClientPort -> Port

File: test/Microsoft.AspNetCore.WebSockets.Test/WebSocketMiddlewareTests.cs [hunk 3]
```diff
@@ -23,10 +24,13 @@ namespace Microsoft.AspNetCore.WebSockets.Test
...
+        private readonly string ClientAddress;
```
Comment (Tratcher) [inline test/Microsoft.AspNetCore.WebSockets.Test/WebSocketMiddlewareTests.cs]: Address.

**Comment by JunTaoLuo** [comment]
@anurse can I merge despite the travis osx failure? I think it's unrelated to my change and I think we can fix it separately if we need.

**Comment by anurse** [comment]
Did https://github.com/aspnet/WebSockets/pull/273 not resolve this issue?

**Comment by JunTaoLuo** [comment]
That one is for 1.1 only and I wasn't too concerned with keeping the tests disabled since 1.1 is EOL'ing next year. This one is 2.1 and I think we should try to fix the tests as much as possible.

**Pull request closed without merging**